File: java/christmas/src/main/java/christmas/Controller/Controller.java
package christmas.Controller;

import christmas.View.*;
import christmas.Model.*;

public class Controller {
    InputView inputView = new InputView();
    OutputView outputView = new OutputView();
    ErrorView errorView = new ErrorView();
    boolean errorTrigger = false;

    public void run() {
        int visitDate = validVisitDate();
        String orderString = validOrderString();
        if(errorTrigger){return;}

        Order order = new Order(orderString);
        if(order.getOrderErrorTrigger()){return;}
        printOrders(orderString,visitDate,order);

        Discount discount = new Discount(order,visitDate);
        printDiscounts(order,discount);

    }
    private void printOrders(String orderString, int visitDate, Order order) {
        OutputView.printDate(visitDate);
        OutputView.printOrder(order);
        OutputView.printTotalPrice(order);
        OutputView.printFreebie(order);

    }
    private void printDiscounts(Order order, Discount discount) {
        OutputView.printDiscount(discount);
        OutputView.printTotalDiscount(discount);
        OutputView.printAfterDCprice(order,discount);
        OutputView.printBadge(discount);
    }

    private String validOrderString() {
        try {
            String menuString = InputView.readMenu();
            String regex = "^([\\p{L}\\d]+-\\d+,)*[\\p{L}\\d]+-\\d+$";
            String menuStringT = menuString.trim();
            boolean isValidMenu = menuStringT.matches(regex);
            if (!isValidMenu) {
                errorView.not_valid_order();
                errorTrigger = true;
            }
            return menuStringT;
        } catch (Exception e) {
            errorView.not_valid_order();
            errorTrigger = true;
            return "";
        }
    }

    private int validVisitDate() {
        try {
            int visitDate = Integer.parseInt(inputView.readDate());
            if (visitDate < 1 || visitDate > 31) {
                errorView.not_valid_date();
                errorTrigger = true;
            }
            return visitDate;
        } catch (Exception e) {
            errorView.not_valid_date();
            errorTrigger = true;
            return 0;
        }
    }
}

File: kotlin/christmas/src/main/kotlin/christmas/controller/controller.kt
package christmas.controller

import christmas.model.discount
import christmas.model.order
import christmas.view.InputView
import christmas.view.ErrorView
import christmas.view.OutputView

class controller {
    private val inputView = InputView()
    private val outputView = OutputView()
    private val errorView = ErrorView()
    private var errorTrigger = false
    fun run() {
        val visitDate = validVisitDate()
        val orderString = validOrderString()
        if (errorTrigger) return

        val order = order(orderString)
        if (order.getOrderErrorTrigger()) return
        printOrders(orderString,visitDate,order)

        val discount = discount(order,visitDate)
        printDiscounts(order,discount)

    }
    private fun printOrders(orderString: String, visitDate: Int, order: order) {
        outputView.printDate(visitDate)
        outputView.printOrder(order)
        outputView.printTotalPrice(order)
        outputView.printFreebie(order)
    }
    private fun printDiscounts(order: order, discount: discount) {
        outputView.printDiscount(discount)
        outputView.printTotalDiscount(discount)
        outputView.printAfterDCprice(order,discount)
        outputView.printBadge(discount)
    }

    private fun validVisitDate(): Int {
        return try {
            val visitDate = Integer.parseInt(inputView.readDate())
            if (visitDate < 1 || visitDate > 31) {
                errorView.notValidDate()
                errorTrigger = true
            }
            visitDate
        } catch (e: Exception) {
            errorView.notValidDate()
            errorTrigger = true
            0
        }
    }
    private fun validOrderString(): String {
        return try {
            val menuString = inputView.readMenu()
            val regex = "^([\\p{L}\\d]+-\\d+,)*[\\p{L}\\d]+-\\d+$"
            val menuStringT = menuString.trim()
            val isValidMenu = menuStringT.matches(Regex(regex))
            if (!isValidMenu) {
                errorView.notValidOrder()
                errorTrigger = true
            }
            menuStringT
        } catch (e: Exception) {
            errorView.notValidOrder()
            errorTrigger = true
            ""
        }
    }
}